NitroxServer/GameLogic/Bases/BuildingManager.cs

```diff
@@ -0,0 +1,318 @@
+        {
+            Log.Error($"Trying to add a ghost to Global Root but another entity with the same id already exists (GhostId: {ghostEntity.Id})");
+            return false;
+        }
+
+        worldEntityManager.AddGlobalRootEntity(ghostEntity);
+        Log.Debug($"Added a GhostEntity: {ghostEntity.Id}");
+        return true;
+    }
+
+    public bool AddModule(PlaceModule placeModule)
+    {
+        ModuleEntity moduleEntity = placeModule.ModuleEntity;
+        if (moduleEntity.ParentId != null)
+        {
+            if (entityRegistry.TryGetEntityById(moduleEntity.ParentId, out Entity parentEntity))
+            {
+                if (parentEntity is not BuildEntity && parentEntity is not VehicleWorldEntity)
+                {
+                    Log.Error($"Trying to add a module to an entity that is not a building/vehicle (ParentId: {moduleEntity.ParentId})");
+                    return false;
+                }
+                if (parentEntity.ChildEntities.Any(childEntity => childEntity.Id.Equals(moduleEntity.Id)))
+                {
+                    Log.Error($"Trying to add a module to a building but another child with the same id already exists (ModuleId: {moduleEntity.Id})");
+                    return false;
+                }
+                parentEntity.ChildEntities.Add(moduleEntity);
+                worldEntityManager.AddGlobalRootEntity(moduleEntity);
+                return true;
+            }
+            Log.Error($"Trying to add a module to a build that isn't registered (ParentId: {moduleEntity.ParentId})");
+            return false;
+        }
+
+        if (entityRegistry.GetEntityById(moduleEntity.Id).HasValue)
+        {
+            Log.Error($"Trying to add a module to Global Root but another entity with the same id already exists ({moduleEntity.Id})");
+            return false;
+        }
+
+        worldEntityManager.AddGlobalRootEntity(moduleEntity);
+        return true;
+    }
+
+    public bool ModifyConstructedAmount(ModifyConstructedAmount modifyConstructedAmount)
+    {
+        if (entityRegistry.TryGetEntityById(modifyConstructedAmount.GhostId, out Entity entity))
+        {
+            if (entity is not GhostEntity && entity is not ModuleEntity)
+            {
+                Log.Error($"Trying to modify the constructed amount of an entity that is not a ghost (Id: {modifyConstructedAmount.GhostId})");
+                return false;
+            }
+            if (modifyConstructedAmount.ConstructedAmount == 0f)
+            {
+                worldEntityManager.RemoveGlobalRootEntity(entity.Id);
+                return true;
+            }
+            if (entity is GhostEntity ghostEntity)
+            {
+                ghostEntity.ConstructedAmount = modifyConstructedAmount.ConstructedAmount;
+            }
+            else if (entity is ModuleEntity moduleEntity)
+            {
+                moduleEntity.ConstructedAmount = modifyConstructedAmount.ConstructedAmount;
```

Comment:
Modules don't ever seem to get finished as the `ConstructedAmount` reaches > 0.9xf but the client acknowledges 100% built so I'm guessing it no longer sends more `ModifyConstructedAmount` packets. And on the server side, the module is not built because it still requires construction.

Tested yesterday on multiple different worlds + different players.

Reply:
Can you confirm the issue is gone ? Supposedly fixed by [5c8dc2e](https://github.com/SubnauticaNitrox/Nitrox/pull/1952/commits/5c8dc2e6ab09e3879ecba34d4c435e703bcfbe75)